[Lee Yong Ler] iP

yongler opened the pull request:
### DukePro [](###dukepro)

> “Your mind is for having ideas, not holding them.” – David Allen [(source)](https://dansilvestre.com/productivity-quotes)

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2122s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

----

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```Java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

yongler committed:
Level-1

yongler committed:
Level-2

yongler committed:
Level-3

yongler committed:
Level-4

yongler committed:
A-TextUiTesting

yongler committed:
A-TextUiTesting

yongler committed:
Level-6

yongler committed:
branch-Level-8

yongler committed:
branch-Level-7

yongler committed:
Merge branch-Level-8

yongler committed:
Level-7

yongler committed:
Level-8

yongler committed:
A-MoreOOP

yongler committed:
A-Packages

yongler committed:
A-JUnit

yongler committed:
A-Jar

yongler committed:
A-CodingStandard

yongler committed:
Level-9

yongler committed:
A-JavaDoc

yongler committed:
Merge branch 'A-CodingStandard'

yongler committed:
Merge branch 'Level-9'

yongler committed:
Merge branch 'A-JavaDoc'

yongler committed:
Fix minor bugs

ryantanlien commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,160 @@
+            String stuff = "";
+            String deadline = "";
+
+            for (int i =1;i< str.length;i++) {
```
Perhaps the for loops here could be better formatted for readability? For instance, more spaces could be added between operators and operands. (Line 75 to Line 80)

ryantanlien commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,160 @@
+     * @throws TaskException
+     */
+    private void addTask(String whole_str, String[] str) throws TaskException {
+        if (str[0].compareTo("deadline") == 0) {
```
The complex expressions in the series of if statements make the code a little more difficult to read. Generic parameter names as well. Perhaps this could be refactored for better readability?

ryantanlien commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,160 @@
+     * @param str the whole input string, split.
+     * @throws TaskException
+     */
+    private void addTask(String whole_str, String[] str) throws TaskException {
```
Don't you think this method is doing too many things at once? It handles input and then handles adding to the task list. Maybe these responsibilities can be further separated into classes and methods? (SRP - Single Responsibility Principle and KISS - Keep it Simple Stupid!)

One suggestion is to move adding task responsibilities to the TaskList class and handle inputs in a new class.

ryantanlien commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,68 @@
+     *
+     * @return [type] where type corresponds to task type.
+     */
+    public String getSymbol() {
```
Would it be better if getSymbol is an abstract method implemented by the other Task sub-classes?

ryantanlien reviewed (commented):
I like your implementation for handling exceptions! Especially the part where it throws a default error message based on the type of task. Maybe I could use that for my own code. 

Have a read through some of my comments, hopefully they provide some value to you! 😄

joshuayapwj98 commented on src/main/java/duke/exceptions/TaskException.java:
```diff
@@ -0,0 +1,19 @@
+package duke.exceptions;
+
+public class TaskException extends DukeException {
+    public static final String TASK_ERROR = "☹ OOPS!!! I'm sorry, but I don't know what that means :-(";
+    public static final String TODO_ERROR = "☹ OOPS!!! The description of a todo cannot be empty.";
+    public static final String EVENT_ERROR = "☹ OOPS!!! The description of a event cannot be empty.";
+    public static final String DEADLINE_ERROR = "☹ OOPS!!! The description of a deadline cannot be empty.";
+
```
Very elegant to make these strings final 😊

joshuayapwj98 commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,160 @@
+
+            for (int i =1;i< str.length;i++) {
+                if (str[i].compareTo("/at") == 0) {
+                    for (int j =1;j<i;j++) {
+                        stuff += str[j] + " ";
+                    }
+                    for (int j =i+1;j<str.length;j++) {
+                        deadline += str[j] + " ";
+                    }
+                    break;
+                }
```
A neat set of for-loops!

However, just a suggestion:
```suggestion
for (int j  = 1; j < i; j++) {
    stuff += str[j] + " ";
}
for (int j = i + 1; j < str.length; j++) {
    deadline += str[j] + " ";
}
break;

joshuayapwj98 commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,160 @@
+                for (int i = 0; i < tasks.size(); i++) {
+                    System.out.printf("%d. %s\n", i + 1, tasks.get(i));
+                }
+            } else if (str[0].compareTo("mark") == 0) {
+                int k = Integer.parseInt(str[1]);
+                tasks.get(k - 1).setAsDone();
+                System.out.println("Nice! I've marked this duke.task as done: ");
+                System.out.println(tasks.get(k - 1));
+            } else if (str[0].compareTo("unmark") == 0) {
+                int k = Integer.parseInt(str[1]);
+                tasks.get(k - 1).setAsUndone();
+                System.out.println("OK, I've marked this duke.task as not done yet:");
+                System.out.println(tasks.get(k - 1));
+            } else if (str[0].compareTo("delete") == 0) {
+                int k = Integer.parseInt(str[1]);
+                System.out.println("Noted. I've removed this duke.task: ");
+                System.out.println(tasks.get(tasks.size()-1).toString());
+                tasks.remove(k-1);
+                System.out.printf("Now you have %d tasks in the list.\n", tasks.size());
+            } else if (str[0].compareTo("find") == 0) {
+                String toFind = str[1];
+
+                int count = 1;
+                System.out.println("Here are the matching tasks in your list:");
+                for (Task task: tasks) {
+                    if (task.getDescription().contains(toFind)) {
+                        System.out.printf("%d. %s\n", count, task);
+                    }
+                }
```
Comprehensive if-else statements!

However, just a minor suggestion. The anonymous variables declared inside each if-else and the different system println makes it hard to read the code 💫  Perhaps some extra indentation or shifting the code out of this if-else is more appropriate?

Overall, still good work 😄

joshuayapwj98 reviewed (commented):
Overall, a very elegant and neat programme!

Accounts for the different use/edge cases that the programme might have! Just some small pointers about the indentations, whitespace and maybe shifting of some print line code.

Nonetheless, good job and keep up the progress!!!

yongler committed:
Stash

yongler committed:
Merge branch 'add-gradle-support'

yongler committed:
A-Gradle

redpelican2108 commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,166 @@
+     * @throws TaskException if no details when adding.
+     */
+    private void addTask(String[] str) throws TaskException {
+        if (str[0].compareTo("deadline") == 0) {
```
The code you have uses deep nesting code. Perhaps you could use a switch case, and abstract out the if else clauses for better readability? 
I've noticed this in other parts of your code as well.

redpelican2108 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,73 @@
+public class Task {
+    protected String description;
+    protected boolean isDone;
+    protected TaskType taskType;
```
Tasktype is a redundant variable as it is already differentiated by the inheriting of the ToDo, Event and Deadline classes. Perhaps you could use method overriding in the subclasses ?

redpelican2108 commented on src/main/java/duke/main/Storage.java:
```diff
@@ -0,0 +1,89 @@
+     * @param str Formatted string in backup file.
+     * @return A task instance.
+     */
+    public Task loadTask(String[] str) {
```
The loadTask method and load method have similar meanings. It appears that the loadTask method is only called by the load method alone. Perhaps you could make the loadTask method private.

redpelican2108 commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,166 @@
+            throw new TaskException("tasks");
+        }
+
+        System.out.println("Got it. I've added this duke.task: ");
```
There are a lot of println in the Duke class, perhaps output can be handled by the UI class?

redpelican2108 reviewed (commented):
I like how the code is clean and well documented. Method and variable names are clear and concise too. Keep up the good work

yongler committed:
Level-10

yongler committed:
Add branch-A-Assertions

Assertion now added to Duke.

Added to learn more about assertions.

Add a line assert false in Duke main method.

yongler committed:
Merge pull request #1 from yongler/branch-A-Assertions

Add branch-A-Assertions

yongler committed:
Stash changes in master

yongler committed:
Merge branch 'master' of https://github.com/yongler/ip

yongler committed:
branch-A-CodeQuality

yongler committed:
branch-A-CodeQuality

yongler committed:
branch-A-CodeQuality

yongler committed:
Stash changes

yongler committed:
Try branch-A-CodeQuality

yongler committed:
Try branch-A-CodeQuality

yongler committed:
Merge pull request #2 from yongler/branch-A-CodeQuality

Branch a code quality

yongler committed:
Merge branch 'master' of https://github.com/yongler/ip

yongler committed:
branch-C-Sort

yongler committed:
Stash changes

yongler committed:
Merge branch 'branch-C-sort'

yongler committed:
Add C-sort extension to SuperDuke

Current implementation of Tasks has deadline along with it,
but no sorting methods

Sorting is needed to enable users to view most important task.

'sort' command was added to sort existing list of tasks.

yongler committed:
Set theme jekyll-theme-cayman

yongler committed:
A-UserGuide

yongler committed:
Merge branch 'master' of https://github.com/yongler/ip

yongler committed:
Change UI.png to Ui.png

yongler committed:
Update picture

yongler committed:
Make final changes
